vital/tests/test_metadata.cxx

```diff
@@ -0,0 +1,123 @@
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <vital/types/metadata.h>
+#include <vital/types/metadata_traits.h>
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <stdint.h>
+
+using namespace ::kwiver::vital;
+
+// ----------------------------------------------------------------------------
+int
+main( int argc, char** argv )
+{
+  ::testing::InitGoogleTest( &argc, argv );
+  return RUN_ALL_TESTS();
+}
+
+// ----------------------------------------------------------------------------
+TEST ( metadata, typed_metadata )
+{
+  // create item
+  std::unique_ptr< typed_metadata > tmds = new typed_metadata( VITAL_META_METADATA_ORIGIN, "item data" );
+  auto* tmdd = new typed_metadata< VITAL_META_PLATFORM_HEADING_ANGLE, double >(
+    std::string( "test double item" ), 3.14159 );
+  auto* tmdi = new typed_metadata< VITAL_META_UNIX_TIMESTAMP, uint64_t >(
+    std::string( "test uint item" ), uint64_t( 314159 ) );
+
+  // test API
+  EXPECT_TRUE( tmds->has_string() );
+  EXPECT_FALSE( tmds->has_double() );
+  EXPECT_FALSE( tmds->has_uint64() );
+  EXPECT_EQ( "item data", tmds->as_string() );
+
+  EXPECT_FALSE( tmdd->has_string() );
+  EXPECT_TRUE( tmdd->has_double() );
+  EXPECT_FALSE( tmdd->has_uint64() );
+  EXPECT_FLOAT_EQ( 3.14159, tmdd->as_double() );
+  EXPECT_EQ( "3.14159", tmdd->as_string() );
+
+  EXPECT_FALSE( tmdi->has_string() );
+  EXPECT_FALSE( tmdi->has_double() );
+  EXPECT_TRUE( tmdi->has_uint64() );
+  EXPECT_EQ( 314159, tmdi->as_uint64() );
+  EXPECT_EQ( "314159", tmdi->as_string() );
+}
+
+// ----------------------------------------------------------------------------
+TEST ( metadata, add_metadata )
+{
+  // create item
+  auto* tmdd = new typed_metadata< VITAL_META_PLATFORM_HEADING_ANGLE, double >(
+    std::string( "test double item" ), 3.14159 );
+  auto rmdi = std::make_shared< typed_metadata< VITAL_META_UNIX_TIMESTAMP, uint64_t > >(
+    std::string( "test uint item" ), uint64_t( 314159 ) );
+
+  metadata meta_collection;
+
+  meta_collection.add< VITAL_META_METADATA_ORIGIN >( "item data" );
+  meta_collection.add( tmdd );
```

Comment:
Are you sure this compiles? I would expect it does not, due to `unique_ptr`'s ctor being `explicit`.